packages/utils/tool-utils/src/snapshotComparisonUtil.ts

```diff
@@ -0,0 +1,131 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { Uint8ArrayToString } from "@fluidframework/common-utils";
+import {
+    ITree,
+    IBlob,
+    TreeEntry,
+    IAttachment,
+} from "@fluidframework/protocol-definitions";
+
+/**
+ * Function that deep sorts an array. It handles cases where array elements are objects or arrays.
+ */
+function deepSortArray(array: any[]) {
+    // Recursively sort arrays, if any, in the array's elements.
+    for (const element of array) {
+        if (element instanceof Array) {
+            deepSortArray(element);
+        } else if (element instanceof Object) {
+            deepSortArraysInObject(element);
+        }
+    }
+
+    // Now that all the arrays in this array's elements have been sorted, sort it by comparing each
+    // element's stringified version.
+    const sortFn = (elem1: any, elem2: any) => {
+        const serializedElem1 = JSON.stringify(elem1);
+        const serializedElem2 = JSON.stringify(elem2);
+        return serializedElem1.localeCompare(serializedElem2);
+    };
+    array.sort(sortFn);
+}
+
+/**
+ * Function that recursively sorts arrays in an object. It handles cases where object elements are arrays
+ * or objects.
+ */
+function deepSortArraysInObject(obj: any) {
+    Object.values(obj).forEach((value) => {
+        if (value instanceof Array) {
+            deepSortArray(value);
+        } else if (value instanceof Object) {
+            deepSortArraysInObject(value);
+        }
+    });
+}
+
+/**
+ * Function that sorts a blob's content. If the content is an object or an array, deep sort arrays in it.
+ */
+function getSortedBlobContent(blobContent: string) {
+    let sortedContent: any;
+    // Try to parse the blob contents in case it's an object or an array.
+    try {
+        const contentObj = JSON.parse(blobContent);
+        if (contentObj instanceof Array) {
+            deepSortArray(contentObj);
+        } else if (contentObj instanceof Object) {
+            deepSortArraysInObject(contentObj);
+        }
+        sortedContent = contentObj;
+    } catch (error) {
+        sortedContent = blobContent;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+    return sortedContent;
+}
+
+/**
+ * Function that deep compares two blobs. It deep sorts arrays in the content, if any, before comparison.
+ */
+function deepCompareBlobs(actualBlob: IBlob, expectedBlob: IBlob) {
+    const content1 = typeof actualBlob.contents === "string"
+        ? actualBlob.contents
+        : Uint8ArrayToString(actualBlob.contents, "base64");
```

Comment:
I think this will fail with image blobs - they are binary blobs, not strings encoded into utf-8.
Though I think the whole process fail (i.e. without your changes) as we do not know how to incorporate blobs that are written out separately outside of snapshot upload (i.e. they behave more like commits)

I think fundamentally, we can't make an assumption here on how to interpret a blob. It's opaque to us. It can be binary data, maybe json payload, maybe something else.

So I worry about this direction. I think better direction is the one where we can mark certain paths / blobs as special that need sorting, vs. try to impose same requirements to whole file - this may potentially miss bugs (where order actually matters), but also can explode as we assume we can treat data as json when it's not.

While not ideal, I think we should also consider existing approach of shipping code writing out data sorted when generating full snapshot - that's how we solve that same problem in other places till now. It's con is that there is shipping code that is needed for testing. But if this code is tiny and runs only in workflows that are driven from test, then maybe it's smaller evil to have?

Reply:
To clarify - today images are TreeEntry.Attachment, so we will not hit that problem. But as a general concept, TreeEntry.Blob can have any data, including binary data.

Reply:
Yeah, sorting can be a good option in the code but if there is some state that tells whether we are in testing vs prod. full snapshot is not a strong enough indication. I was talking to Tony and he mentioned there are options in the Loader that we might be able to use for this. Is there any other global state that we can look at and decide, rather than using parameters passed to an API?
I believe this will be helpful in other similar cases as well where we might want to do some things differently for testing.